twoliter/embedded/imghelper

```diff
@@ -1,6 +1,140 @@
 #!/usr/bin/env bash
-
 # shellcheck disable=SC2034 # variables are used through parameter substitution
+
+DATA_IMAGE_SUFFIX="-data"
+BOOT_IMAGE_SUFFIX="-boot"
+VERITY_IMAGE_SUFFIX="-root"
+ROOT_IMAGE_SUFFIX="-root"
+
+set_image_names() {
+  local filename_prefix
+  local -n os_image_name data_image_name
+  local -n boot_image_name verity_image_name root_image_name
+  filename_prefix="${1:?}"
+  os_image_name="${2:?}"
+  data_image_name="${3:?}"
+  boot_image_name="${4:?}"
+  verity_image_name="${5:?}"
+  root_image_name="${6:?}"
+
+  os_image_name="${filename_prefix}"
+  data_image_name="${filename_prefix}${DATA_IMAGE_SUFFIX}"
+  boot_image_name="${filename_prefix}${BOOT_IMAGE_SUFFIX}"
+  verity_image_name="${filename_prefix}${VERITY_IMAGE_SUFFIX}"
+  root_image_name="${filename_prefix}${ROOT_IMAGE_SUFFIX}"
+}
+
+sanity_checks() {
+  local output_fmt partition_plan ovf_template uefi_secure_boot
+  output_fmt="${1:?}"
+  partition_plan="${2:?}"
+  ovf_template="${3:?}"
+  uefi_secure_boot="${4:?}"
+  case "${output_fmt}" in
+  raw | qcow2 | vmdk) ;;
+  *)
+    echo "unexpected image output format '${output_fmt}'" >&2
+    exit 1
+    ;;
+  esac
+
+  case "${partition_plan}" in
+  split | unified) ;;
+  *)
+    echo "unexpected partition plan '${partition_plan}'" >&2
+    exit 1
+    ;;
+  esac
+
+  # Fail fast if the OVF template doesn't exist, or doesn't match the layout.
+  if [ "${output_fmt}" == "vmdk" ]; then
+    if [ ! -s "${ovf_template}" ]; then
+      echo "required OVF template not found: ${ovf_template}" >&2
+      exit 1
+    fi
+
+    if [ "${partition_plan}" == "split" ]; then
+      if ! grep -Fq '{{DATA_DISK}}' "${ovf_template}"; then
+        echo "Missing data disk in OVF template, which is required for 'split' layout." >&2
+        exit 1
+      fi
+    fi
+
+    if [ "${partition_plan}" == "unified" ]; then
+      if grep -Fq '{{DATA_DISK}}' "${ovf_template}"; then
+        echo "Data disk included in OVF template, which is not supported for 'unified' layout." >&2
+        exit 1
+      fi
+    fi
+
+    if [ "${uefi_secure_boot}" == "yes" ]; then
+      if ! grep -Fq '{{DB_CERT_DER_HEX}}' "${ovf_template}"; then
+        echo "Missing CA certificate field in OVF template, which is required for Secure Boot support." >&2
+        exit 1
+      fi
+    fi
+  fi
+}
+
+decompress_image() {
+  local fmt input_dir filename_prefix output_image ext
```

Comment:
I still think you need to align this with the `compress_image` signature.

Compress:
```
compress_image "vmdk" "os_image" "${OUTPUT_DIR}"
symlink_image "vmdk" "os_image" "${SYMLINK_PREFIX}" "${VERSION_ID}" "${OUTPUT_DIR}"
if [ -s "${DATA_IMAGE}" ] ; then
  compress_image "vmdk" "data_image" "${OUTPUT_DIR}"
  symlink_image "vmdk" "data_image" "${SYMLINK_PREFIX}" "${VERSION_ID}" "${OUTPUT_DIR}"
fi
```

Decompress:
```
decompress_image "vmdk" "os_image" "${OUTPUT_DIR}"
symlink_image "vmdk" "os_image" "${SYMLINK_PREFIX}" "${VERSION_ID}" "${OUTPUT_DIR}"
if [ -s "${DATA_IMAGE}" ] ; then
  decompress_image "vmdk" "data_image" "${OUTPUT_DIR}"
  symlink_image "vmdk" "data_image" "${SYMLINK_PREFIX}" "${VERSION_ID}" "${OUTPUT_DIR}"
fi
```

For that to work, you will want to leave the compressed artifact alone, so that the second call can extract the data image for the split VMDK case. But for consistency you should also not remove it in the raw or qcow2 cases.